File: program_control/sources/assumption.2.c
/*
* @@name:       assumption.2
* @@type:       C
* @@operation:  compile
* @@expect:     success
* @@version:    omp_6.0
*/
#include <stdio.h>
#define N 5

void init(int *arr, int len);
int element_transform(int a);

int main() {
  int arr[N], arr_bang[N];

//Case 1: Use in sequential code
  #pragma omp assume no_openmp
  {   
    init(arr,N);  
  }   
    
//Case 2: Use inside openmp construct
  #pragma omp target teams loop map(to: arr)  map(from: arr_bang)
  for(int i = 0; i < N; i++) {
    #pragma omp assume no_parallelism 
    {   
       arr_bang[i] = element_transform(arr[i]);
    } 
  }
  printf("%d, %d\n", arr_bang[0], arr_bang[N-1]);
    
  return 0;
}

File: program_control/sources/assumption.2.f90
! @@name:       assumption.2
! @@type:       F-free
! @@operation:  compile
! @@expect:     success
! @@version:    omp_6.0

module mm
  interface
    subroutine init(arr, n)
      integer :: arr(*)
      integer :: n
    end subroutine
    function element_transform(a) result(r)
      !$omp declare target
      integer :: a, r
    end function
  end interface
end module

program main
   use mm
   integer, parameter :: N=5
   integer  :: arr(N), arr_bang(N)

!!Case 1: Use in sequential code
   !$omp assume no_openmp
     call init(arr,N)
   !$omp end assume
    
!!Case 2: Use inside openmp construct
   !$omp target teams loop map(to: arr) map(from: arr_bang)
   do i=1,N
      !$omp assume no_parallelism
        arr_bang(i) = element_transform(arr(i))
      !$omp end assume
   enddo

   print *, arr_bang(1), arr_bang(N)

end program main